### ISSUE TITLE
Can Random Forests overfit?

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10828]
Can Random Forests overfit?<br>
<br>
https://ift.tt/wmZNA6P<br>
<br>
<br>
<br>

<div><div><div><div>
Can Random Forests overfit?
<hr>
<h1>Can Random Forests overfit?</h1>
<i></i><p>December 26, 2022 · <span title="Estimated read time">
7 mins · 1430 words
</span>
</p> <h1>Introduction</h1>
<p>If you’re like me - a DS/MLE with practical experience - you may be surprised by the title of this post. At first, I assumed the answer was a clear “yes.” However, when someone with more experience asked me this question, I realized that the answer might not be as obvious as I thought. And after some googling I found these quotes by Leo Breiman, the creator of the Random Forest algorithm</p>
<blockquote>
<p>Random forests does not overfit. You can run as many trees as you want.<br><em>(<a href="https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm#remarks">website</a>)</em></p>
</blockquote>
<p>and</p>
<blockquote>
<p>Random forests do not overfit as more trees are added, but produce a limiting value of the generalization error.<br><em>(<a href="https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf">paper</a>)</em></p>
</blockquote>
<p>Can random forests really not overfit? It seems counterintuitive, but in this post I will explain with math and experiments why it is possible under certain conditions.</p>
<p>To write this post I’ve used this <a href="https://arxiv.org/pdf/1407.7502.pdf">PhD dissertation</a> by <a href="https://glouppe.github.io/">Gilles Loupe</a> (one of the creators of sklearn), this <a href="https://mljar.com/blog/random-forest-overfitting/">blog post</a> that studies the same question, and the original <a href="https://www.stat.berkeley.edu/~breiman/randomforest2001.pdf">paper</a> by Leo Breiman.</p>
<h1>Maths</h1>
<h2>Model definition</h2>
<p>Let’s start by defining mathematically what we mean by a random forest. Let’s assume that we already know how to create single classification trees. Let’s call $D$ the dataset, a decision tree $T$, and $\Theta$ the hyperparameters of the tree. The model generated by these pieces is then $\Psi_{D, \Theta} = T(D, \Theta)$. For the sake of simplicity, let’s assume that all hyperparameters are fixed and that the only free hyperparameter is the random seed, ie: the values <code>max_depth</code> , <code>min_sample_split</code> , <code>min_sample_leaf</code> , etc. are fixed.</p>
<p>Random forests are a combination of tree predictors such that each tree depends on the values of a random vector sampled independently and with the same distribution for all trees in the forest. Therefore, for a set of $M$ randomized trees $\{ \Psi_{D, \Theta_i} | i = 1, …, M \}$ that learn from the same dataset $D$we define a random forest model as <sup></sup></p>
\[\Phi_{D, \Theta_1, ..., \Theta_M}(x) = \frac{1}{M} \sum_i \Psi_{D, \Theta_i}(x)\]
<h2>Bias-Variance decomposition</h2>
<p>Now that we know how to build a random forest let’s study its bias and variance, which are the key ingredients of overfiting. For a pair $x$ and $y$, where $x$ is the vector of features, and $y$ is the target value, let’s define for a single tree $\Psi_{D, \Theta}$</p>
\[\mu_{D, \Theta_i}(x) = \mathbb{E}_{D, \Theta_i} \left[\Psi_{D, \Theta_i}(x)\right]\]
<p>and</p>
\[\sigma^2_{D, \Theta_i}(x) = \mathbb{V}_{D, \Theta_i} \left[\Psi_{D, \Theta_i}(x)\right]\]
<p>Also, we know that the expected generalization error of a model decomposes as</p>
\[\text{error}(x) = \text{bias}^2(x) + \text{variance}(x) + \text{noise}(x)\]
<p>We’re interested in the generalization error that our random forest has depending on its hyperparameters. Let’s start with the bias, which is defined as the difference between the actual value and the model’s expected value</p>
\[\mathbb{E}_{D, \Theta_1, ..., \Theta_M}\left[ \Phi_{D, \Theta_1, ..., \Theta_M}(x) \right] = \mathbb{E} \left[ \frac{1}{M} \sum_i \Psi_{D, \Theta_i}(x) \right] = \frac{1}{M} \sum_i \mathbb{E} \left[\Psi_{D, \Theta_i}(x)\right] = \mu_{D, \Theta}(x)\]
<p>because random variables $\Theta_i$ are independent and follow the same distribution. Therefore</p>
\[\text{bias}(x) = y - \mu_{D, \Theta}(x)\]
<p>which means that the bias of the random forest is the same as the bias of a single tree, so combining decision trees has no effect on the bias.</p>
<p>Let’s look now at what happens with the variance. After some maths (page 66 of <a href="https://arxiv.org/pdf/1407.7502.pdf">this paper</a>) it can be shown that the variance of a random forest of $M$ trees is</p>
\[\text{variance}(x) = \rho(x) \sigma^2_{D, \Theta}(x) + \frac{1 - \rho(x)}{M} \sigma^2_{D, \Theta}(x)\]
<p>where $\rho(x)$ is the Pearson correlation between the predictions of two decision trees.</p>
<p>Therefore, the generalization error is given by</p>
\[\text{error}(x, y; M) = (y - \mu_{D, \Theta}(x))^2 + \rho(x) \sigma^2_{D, \Theta}(x) + \frac{1 - \rho(x)}{M} \sigma^2_{D, \Theta}(x) + \text{noise}(x)\]
<h2>Overfitting</h2>
<p>Now that we know how to compute the generalization error of our model we can study how it behaves when changing the hyperparameters. Since each tree is built using only a bootstrap of the full dataset we have $\rho(x) &lt; 1$, ie: different trees in the random forest give different predictions for the same input. Therefore, if $K &gt; L$ we have</p>
\[\text{error}(K) &lt; \text{error}(L)\]
<p>Also, when $M \to \infty$ we have</p>
\[\text{error}(x, y) = (y - \mu_{D, \Theta}(x))^2 + \rho(x) \sigma^2_{D, \Theta}(x) + \text{noise}(x)\]
<p>As a result, the expected generalization error of the random forest is smaller than the expected error of a single decision tree. This is, as we add more trees to the random forest the generalization error reduces.</p>
<p>Therefore, adding more trees to the model doesn’t make the model more prone to overfitting, in fact it reduces the generalization error! But it doesn’t mean that a random forest can’t overfit. For example, imagine the simplest case of a random forest with only one tree. If we allow the tree to have an infinite depth it can learn almost perfectly the training data (up to collisions), but then it’ll fail to generalize to other data.</p>
<h1>Experiment</h1>
<p>In the above section, we’ve seen mathematically that adding more trees to the model shouldn’t make the model overfit. In this section, we are going to take a more practical approach and show it directly with some experiments.</p>
<p>To study the effect of the number of trees on overfitting we’ll generate a synthetic dataset using the <code>sklearn.datasets.make_regression</code> method.</p>
<div><div><pre><code><span>from</span> <span>sklearn.datasets</span> <span>import</span> <span>make_regression</span>
<span>from</span> <span>sklearn.model_selection</span> <span>import</span> <span>train_test_split</span>

<span>X</span><span>,</span> <span>y</span> <span>=</span> <span>make_regression</span><span>(</span><span>n_samples</span><span>=</span><span>10_000</span><span>,</span> 
                       <span>n_features</span><span>=</span><span>50</span><span>,</span> 
                       <span>n_informative</span><span>=</span><span>30</span><span>,</span> 
                       <span>noise</span><span>=</span><span>10</span><span>)</span>
<span>X_train</span><span>,</span> <span>X_test</span><span>,</span> <span>y_train</span><span>,</span> <span>y_test</span> <span>=</span> <span>train_test_split</span><span>(</span><span>X</span><span>,</span> <span>y</span><span>,</span> 
                                                    <span>test_size</span><span>=</span><span>0.3</span><span>,</span>
                                                    <span>random_state</span><span>=</span><span>42</span><span>)</span>
</code></pre></div></div>
<p>Now we can train multiple random forests, each one with a different number of trees.</p>
<div><div><pre><code><span>from</span> <span>sklearn.ensemble</span> <span>import</span> <span>RandomForestRegressor</span>
<span>from</span> <span>sklearn.metrics</span> <span>import</span> <span>mean_squared_error</span>

<span>rf</span> <span>=</span> <span>RandomForestRegressor</span><span>()</span>
<span>n_estimators</span> <span>=</span> <span>[]</span>
<span>train_mse</span> <span>=</span> <span>[]</span>
<span>test_mse</span> <span>=</span> <span>[]</span>
<span>for</span> <span>n</span> <span>in</span> <span>range</span><span>(</span><span>1</span><span>,</span> <span>50</span><span>,</span> <span>1</span><span>):</span>
  <span>rf</span><span>.</span><span>n_estimators</span> <span>=</span> <span>n</span>
  <span>rf</span><span>.</span><span>fit</span><span>(</span><span>X_train</span><span>,</span> <span>y_train</span><span>)</span>
  <span>y_train_predicted</span> <span>=</span> <span>rf</span><span>.</span><span>predict</span><span>(</span><span>X_train</span><span>)</span>
  <span>y_test_predicted</span> <span>=</span> <span>rf</span><span>.</span><span>predict</span><span>(</span><span>X_test</span><span>)</span>
  <span>mse_train</span> <span>=</span> <span>mean_squared_error</span><span>(</span><span>y_train</span><span>,</span> <span>y_train_predicted</span><span>)</span>
  <span>mse_test</span> <span>=</span> <span>mean_squared_error</span><span>(</span><span>y_test</span><span>,</span> <span>y_test_predicted</span><span>)</span>
  <span>n_estimators</span><span>.</span><span>append</span><span>(</span><span>n</span><span>)</span>
  <span>train_mse</span><span>.</span><span>append</span><span>(</span><span>mse_train</span><span>)</span>
  <span>test_mse</span><span>.</span><span>append</span><span>(</span><span>mse_test</span><span>)</span>
</code></pre></div></div>
<p>And we can finally plot the MSE for both train and test datasets.</p>
<p><img alt="figure-1" src="https://www.alexmolas.com/docs/random-forest-overfit/mse-vs-estimators.svg"></p>
<p>As you can see, there’s a gap between the train and test MSE which means that there’s some overfitting. However, as we add more trees the gap between the train and test MSE doesn’t increase, which means that adding more trees doesn’t make the model more prone to overfitting. In fact, as we add more trees to the model the gap between the curves reduces. In fact, in the next plot we see how <code>Gap = Test MSE - Train MSE</code> reduces as <code>n_estimators</code> increases</p>
<p><img alt="figure-2" src="https://www.alexmolas.com/docs/random-forest-overfit/gap-vs-estimators.svg"></p>
<p>Notice also that the gap between the train and test curves — aka overfitting — can be changed by tuning other hyperparameters, such as <code>max_depth</code> . In the next plot I show how the gap changes depending on <code>max_depth</code> and <code>n_estimators</code>.</p>
<p><img alt="figure-2" src="https://www.alexmolas.com/docs/random-forest-overfit/gap-vs-max-depth.svg"></p>
<h1>Conclusions</h1>
<ul>
<li>Yes, random forests can overfit since a single tree can overfit.</li>
<li>The bias of random forests is the same as the bias of a single tree, however, the variance decreases as we add more trees to the model, and this is where the power of random forests comes from.</li>
<li>Overfitting in a random forest model can be tuned using other hyperparameters such as <code>max_depth</code> , but increasing <code>n_estimators</code> doesn’t increase the gap between train and test performance.</li>
</ul>
<hr>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via  <br>
<br>
February 8, 2024 at 09:47AM
